explicitly reset stderr/stdout in precompilation

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125289

I was seeing a weird bug where after running max-autotune my stdout would be misdirected. other people have not been able to repro this.

cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

eellison committed:
explicitly reset stderr/stdout in precopmilation

[ghstack-poisoned]

shunting314 commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1007,15 +1008,38 @@ def no_op(*args, **kwargs):
+            original_stderr = sys.stderr
+
+            @contextmanager
+            def restore_stdout_stderr():
```
rename to preserve_stdout_stderr and move to utils.py?

shunting314 commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1007,15 +1008,38 @@ def no_op(*args, **kwargs):
+                    a = sys.stdout
+                    sys.stdout = original_stdout
+                    sys.stderr = original_stderr
+                    print(a)
```
remove these debugging code before merging.

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1007,15 +1008,38 @@ def no_op(*args, **kwargs):
+            original_stdout = sys.stdout
+            original_stderr = sys.stderr
+
+            @contextmanager
```
I haven't figured out yet. I added code which would error when stdout is reassigned but it didnt fire. All of the uses of stdout in torch/_inductor/ are unrelated. I also printed out all of the "bad" sys.stdout on exit. The StringIO objects do not match any of the same StringIO objects we construct inside torch inductor. Sometimes the bad objects are `TextIOWrapper` which we dont construct inside torch/_inductor

mlazos reviewed (approved):
Approved, but please file an isssue with the repro so people will know about this if they hit something similar

eellison committed:
Update on "explicitly reset stderr/stdout in precompilation"


I was seeing a weird bug where after running max-autotune my stdout would be misdirected. other people have not been able to repro this.

cc ezyang msaroufim bdhirsh anijain2305 chauhang voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
